doc/slack.8
.\" $Header$
.\" vim:tw=72:filetype=nroff
.\"
.\"       manpage for slack.conf
.\"
.TH slack 8 2004-10-22 "Administrative commands" 

.SH NAME
slack \- Sysadmin's lazy autoconfiguration kit

.SH SYNOPSIS

\fBslack\fR [\fIoption ...\fR] [\fIrole ...\fR]

.SH DESCRIPTION

slack is a master command which coordinates the activities of its
backends, which variously:

.IP \(bu
determine the list of roles to be installed on this server
.IP \(bu
create a local cached copy of the role files from the central repository
.IP \(bu
merge file trees from subroles into a single, unified tree
.IP \(bu
install files onto the local filesystem
.IP \(bu
run scripts before and after installation

.PP
Options you give to slack will be generally passed along to the backends
where relevant.

.SH OPTIONS
.TP
\fB\-h\fR, \fB\-\-help\fR
Print a usage statement.
.TP
\fB\-\-version\fR
Print the version and exit.
.TP
\fB\-v\fR, \fB\-\-verbose\fR
Increase verbosity.  Can be specified multiple times.
.TP
\fB\-\-quiet\fR
Don't be verbose (Overrides previous uses of --verbose).
.TP
\fB\-C\fR, \fB\-\-config FILE\fR
Use the specfied FILE for configuration instead of the default,
.IR /etc/slack.conf .
.TP
\fB\-s\fR, \fB\-\-source DIR\fR
Source directory for slack files
.TP
\fB\-e\fR, \fB\-\-rsh COMMAND\fR
Remote shell for rsync
.TP
\fB\-c\fR, \fB\-\-cache DIR\fR
Local cache directory for slack files
.TP
\fB\-t\fR, \fB\-\-stage DIR\fR
Local staging directory for slack files
.TP
\fB\-r\fR, \fB\-\-root DIR\fR
Root destination for slack files
.TP
\fB\-\-no\-sync\fR
Skip the slack-sync step (useful if you're pushing stuff into the CACHE
outside slack).
.TP
\fB\-\-no\-files\fR
Don't install any files in ROOT, but tell rsync to print what it
would do.
.TP
\fB\-\-no\-scripts\fR
Don't run scripts
.TP
\fB\-n\fR, \fB\-\-dry\-run\fR
Same as \-\-no\-files \-\-no\-scripts  (CACHE, STAGE will still be
updated)
.TP
\fB\-\-role\-list\fR
Role list for
.BR slack-getroles (8).
.TP
\fB\-b\fR, \fB\-\-backup\fR
Make backups of existing files in ROOT that are overwritten.  This
option defaults to on if it is not set to 0 in a config file or disabled
with --nobackup on the command line.
.TP
\fB\-\-backup\-dir\fR
Put backups from the
.B \-\-backup
option into this directory.
.TP
\fB\-H\fR, \fB\-\-hostname HOST\fR
Pretend to be running on HOST, instead of the name given by
gethostname(2).
.TP
\fB\-\-preview MODE\fR
Do a diff of scripts and files before running them.
MODE can be one of 'simple' or 'prompt' (See PREVIEW MODES, below).
.TP
\fB\-\-diff PROG\fR
Use this diff program for previews.
.TP
\fB\-\-sleep TIME\fR
Randomly sleep between 1 and TIME seconds before starting
operations.  Useful in crontabs.

.SH PREVIEW MODES

.PP
Preview functionality is new in slack 0.14.0.  I haven't quite
worked out how things will work, so this usage is somewhat subject to
change in future versions.  I thought I would try it this way and see
how people like it.
.PP
In 'simple' mode, after syncing and staging the files directory, slack
will present a diff of the files and scripts.  In this mode, slack will
not run the preinstall or fixfiles scripts, and because of this, it may
provide some false output about permissions changes to files.
.PP
In 'prompt' mode, after syncing and staging the files directory, slack
will diff the script directory.  If there are differences, slack will
present them to you and ask you if you want to continue.  If you say
no, it will exit.  If you say yes, it will stage the scripts directory,
run the preinstall and fixfiles scripts, and then diff the files in the
stage with those in the root.  If there are differences, slack will
present them to you and ask you if you want to continue.  If you say no,
it will exit.  If you say yes, it will install the files and run the
postinstall script.
.PP
So, the 'simple' mode is easy to use, and will be accurate if you don't
use fixfiles.  The 'prompt' mode will be accurate if you use fixfiles,
but requires some interaction.
.PP
Why can't we just have one mode that works with fixfiles and requires no
interaction?  Well, that would require slack to understand what your
free-form fixfiles executable was going to do, which would either
require some kind of universe simulator or would require you to write
your fixfiles in a less free-form way, which would make slack less like
slack.

.SH EXAMPLES

.PP
To install all the roles configured in the role list for a server:
.RS
slack
.RE
.PP
To install a specific role:
.RS
slack
.I rolename
.RE
.PP
To test a new role before checking in the changes:
.RS
slack --source 
.IR user @ workstation :/home/ user /.../slack
.I rolename
.RE
.PP
To avoid killing your master server when calling from cron:
.RS
slack --sleep 3600
.RE



.SH FILES
.I /etc/slack.conf
.SH SEE ALSO
.BR slack.conf (5),
.BR rsync (1)


ChangeLog
2008-04-20  Alan Sundell  <sundell@gmail.com>
	* Version 0.15.2

2008-04-18  Alan Sundell  <sundell@gmail.com>
	* Makefile: fix 'dist' target to work with svn

2008-03-02  Alan Sundell  <alan@sundell.net>
	* Makefile: remove debian directory from tarball, to make Andrew
	    Pollock happy; make a separate tarball for the (suggested) debian
	    directory, to make me happy until his debian patch becomes
	    available
	* src/Slack.pm, test/01_module.t: apply Steve Atwell's get_system_exit
	    patch.  This uses the POSIX macros for exit statues, makes the
	    tests work on OpenBSD, and removes the bit that checks core dumps,
	    since it was kind of silly and awkward to test, anyway.
	* src/Makefile: another OpenBSD patch from Steve Atwell -- don't
	    specify -p with -d when using install (doesn't make sense, anyway)

2008-01-19  Alan Sundell  <alan@sundell.net>
	* Version 0.15.1

2008-01-18  Alan Sundell  <alan@sundell.net>
	* Makefile: minor changes to 'dist' target
	* src/slack-sync: use --files-from=- to do the whole sync at once. 
	    On my systems for a 15-role sync this cuts the time from
	    4-5 sec to 0.4-0.5 sec.
	* test/*: unittest updates, change handling of temp dir
	* Slack.pm: wrap_rsync() and wrap_rsync_fh() replace some duplicated
	    code in the backends.
	
2008-01-14  Alan Sundell  <alan@sundell.net>
	* Version 0.15.0
	* src/Makefile: install slack.conf without the execute bit (thanks to
	    Andrew Pollock)
	* Makefile, debian/*: pull in a bunch of changes from Andrew Pollock's
	    debian patch (perhaps to his frustration), with a few changes
	    to preserve permissions on /var/lib/slack and /var/cache/slack;
	    and a few of my own changes to the way debs are built.
	* src/slack: Propagate --rsh to backends so the option actually works
	    
2008-01-12  Alan Sundell  <alan@sundell.net>
	* Slack.pm, slack-sync, slack-getroles: patch to add --rsh/-e option
	    from David Lowry.  This allows us to explicitly enable the
	    behavior implied by a double colon in SOURCE.
	* slack-stage: fix from Brian Brazil for a bug that occurs with
	    files in a subrole that have the same size and timestamp as
	    those in a parent role, such as can occur when the two files
	    are checked into a repository at the same time.
	* slack: add a --sleep option for a random sleep.  Useful in
	    crontabs, where doing 'sleep $(($RANDOM % 3600)) && slack'
	    can run you afoul of cron's treatment of '%'.
	* Slack.pm: add a --version option
	* slack.spec: integrate some minor changes from Gordon Messmer's
	    patch (much already done by David Lowry's patch), most
	    significantly adding username to buildroot

2007-06-15  Alan Sundell  <alan@sundell.net>
	* doc/slack.conf.5: describe the four forms of SOURCE

2007-01-25  Alan Sundell  <alan@sundell.net>
	* src/Makefile: mkdir /usr/bin to fix slack-diff install (thanks
			to Robert Flemming and Chris Heiser)
	* slack.spec: new, much-improved spec file from David Lowry

2006-11-05  Alan Sundell  <alan@sundell.net>
	* Version 0.14.1
	* slack-getroles: remove --delete from rsync options (incompatible
	    with newer versions [2.6.8] of rsync without -r); call
	    check_system_exit() properly

2006-10-12  Alan Sundell  <alan@sundell.net>
	* Version 0.14.0
	* slack.8: added missing and new options, and a section
	    on preview modes
	* slack: made --preview require a mode to avoid confusion of preview
	    modes and role names; made 'prompt' preview mode run the
	    preinstall/fixfiles scripts before showing file diff

2006-09-25  Alan Sundell  <alan@sundell.net>
	* slack-rolediff: renamed from slack-preview to avoid name conflict
	* 08_diff.t: tests for slack-diff

2006-09-21  Alan Sundell  <alan@sundell.net>
	* COPYING, GPL, CREDITS: add license, attribution
	* README: lots of new text about slack
	* slack-installfiles: when creating ROOT, make it mode 0755

2006-03-21  Alan Sundell  <alan@sundell.net>
	* slack-diff: new diffing tool
	* slack-preview: new slack backend for diffing tool
	* slack-stage: allow specifying a subdir (files/script) to stage

2005-10-30  Alan Sundell  <alan@sundell.net>
	* slack.spec: expect slack.conf in sysconfdir (since that's where
	    we install it)
	* slack-installfiles: use program name when logging "Creating
	    backup directory" message (like all other messages)

2005-02-09  Alan Sundell  <alan@sundell.net>
	* Version 0.13.2
	* slack-installfiles: fail gracefully if no files to install

2005-01-08  Alan Sundell  <alan@sundell.net>
	* Version 0.13.1
	* test: added tests for each of the backends
	    * testsource: mockup roles for testing
	    * test_util.pm: common functions, variables for testing
	    * gen_config_file: generate config files with full paths
			       for testing
	* test: added tests for the frontend
	* slack: call check_system_exit with full name (fixes undefined
	    subroutine error when a backend failed)
	* Slack.pm: say commands "exited" with their exit code,
	    rather than "returned"
	* slack-stage: in the base role, delete the destination if the
	    source doesn't exist
	* slack-sync: remove unused variable

2005-01-05  Alan Sundell  <alan@sundell.net>
	* slack-getroles: add a --remote-role-list option to force local sync
	* slack-sync: we don't need File::Find anymore
	* slack-runscript: mention when we skip a non-existant script
	    if verbose is on.

2005-01-04  Alan Sundell  <alan@sundell.net>
	* added support for "make dist"
	* slack-getroles: don't prepend source for paths to role-list
	    beginning with './' (important for testing)
	* Slack.pm: get_options(): generate default usage if no usage given
		    get_options(): don't pass back the --quiet handler
	* added support for "make test"
	* test: added Slack.pm tests
	* [Makefiles]: support a common set of install-* targets

2004-12-22  Alan Sundell  <alan@sundell.net>
	* Version 0.13
	* slack,Slack.pm,slack-getroles: add a --hostname option,
	    use that to select roles when provided.

src/Slack.pm
# $Id: Slack.pm 189 2008-04-21 00:52:56Z sundell $
# vim:sw=2
# vim600:fdm=marker
# Copyright (C) 2004-2008 Alan Sundell <alan@sundell.net>
# All Rights Reserved.  This program comes with ABSOLUTELY NO WARRANTY.
# See the file COPYING for details.

package Slack;

require 5.006;
use strict;
use Carp qw(cluck confess croak);
use File::Find;
use POSIX qw(WIFEXITED WEXITSTATUS WIFSIGNALED WTERMSIG);

use base qw(Exporter);
use vars qw($VERSION @EXPORT @EXPORT_OK $DEFAULT_CONFIG_FILE);
$VERSION = '0.15.2';
@EXPORT    = qw();
@EXPORT_OK = qw();

$DEFAULT_CONFIG_FILE = '/etc/slack.conf';

my $term;

my @default_options = (
    'help|h|?',
    'version',
    'verbose|v+',
    'quiet',
    'config|C=s',
    'source|s=s',
    'rsh|e=s',
    'cache|c=s',
    'stage|t=s',
    'root|r=s',
    'dry-run|n',
    'backup|b',
    'backup-dir=s',
    'hostname|H=s',
);

sub default_usage ($) {
  my ($synopsis) = @_;
  return <<EOF;
Usage: $synopsis

Options:
  -h, -?, --help
      Print this help message and exit.

  --version
      Print the version number and exit.

  -v, --verbose
      Be verbose.

  --quiet
      Don't be verbose (Overrides previous uses of --verbose)

  -C, --config  FILE
      Use this config file instead of '$DEFAULT_CONFIG_FILE'.

  -s, --source  DIR
      Source for slack files

  -e, --rsh  COMMAND
      Remote shell for rsync

  -c, --cache  DIR
      Local cache directory for slack files

  -t, --stage  DIR
      Local staging directory for slack files

  -r, --root  DIR
      Root destination for slack files

  -n, --dry-run
      Don't write any files to disk -- just report what would have been done.

  -b, --backup
      Make backups of existing files in ROOT that are overwritten.

  --backup-dir  DIR
      Put backups into this directory.

  -H, --hostname  HOST
      Pretend to be running on HOST, instead of the name given by
        gethostname(2).
EOF
}
# Read options from a config file.  Arguments:
#       file    => config file to read
#       opthash => hashref in which to store the options
#       verbose => whether to be verbose
sub read_config (%) {
  my %arg = @_;
  my ($config_fh);
  local $_;

  confess "Slack::read_config: no config file given"
    if not defined $arg{file};
  $arg{opthash} = {}
    if not defined $arg{opthash};

  open($config_fh, '<', $arg{file})
    or confess "Could not open config file '$arg{file}': $!";

  # Make this into a hash so we can quickly see if we're looking
  # for a particular option
  my %looking_for;
  if (ref $arg{options} eq 'ARRAY') {
    %looking_for = map { $_ => 1 } @{$arg{options}};
  }

  while(<$config_fh>) {
    chomp;
    s/#.*//; # delete comments
    s/\s+$//; # delete trailing spaces
    next if m/^$/; # skip empty lines

    if (m/^[A-Z_]+=\S+/) {
      my ($key, $value) = split(/=/, $_, 2);
      $key =~ tr/A-Z_/a-z-/;
      # Only set options we're looking for
      next if (%looking_for and not $looking_for{$key});
      # Don't set options that are already set
      next if defined $arg{opthash}->{$key};

      $arg{verbose} and print STDERR "Slack::read_config: Setting '$key' to '$value'\n";
      $arg{opthash}->{$key} = $value;
    } else {
      cluck "Slack::read_config: Garbage line '$_' in '$arg{file}' line $. ignored";
    }
  }

  close($config_fh)
    or confess "Slack::read_config: Could not close config file: $!";

  # The verbose option is treated specially in so many places that
  # we need to make sure it's defined.
  $arg{opthash}->{verbose} ||= 0;

  return $arg{opthash};
}

# Just get the exit code from a command that failed.
# croaks if anything weird happened.
sub get_system_exit (@) {
  my @command = @_;

  if (WIFEXITED($?)) {
    my $exit = WEXITSTATUS($?);
    return $exit if $exit;
  }
  if (WIFSIGNALED($?)) {
    my $sig = WTERMSIG($?);
    croak "'@command' caught sig $sig";
  }
  if ($!) {
    croak "Syserr on system '@command': $!";
  }
  croak "Unknown error on '@command'";
}

sub check_system_exit (@) {
  my @command = @_;
  my $exit = get_system_exit(@command);
  # Exit is non-zero if get_system_exit() didn't croak.
  croak "'@command' exited $exit";
}

# get options from the command line and the config file
# Arguments
#       opthash => hashref in which to store options
#       usage   => usage statement
#       required_options => arrayref of options to require -- an exception
#               will be thrown if these options are not defined
#       command_line_hash => store options specified on the command line here
sub get_options {
  my %arg = @_;
  use Getopt::Long;
  Getopt::Long::Configure('bundling');

  if (not defined $arg{opthash}) {
    $arg{opthash} = {};
  }

  if (not defined $arg{usage}) {
    $arg{usage} = default_usage($0);
  }

  my @extra_options = ();  # extra arguments to getoptions
  if (defined $arg{command_line_options}) {
    @extra_options = @{$arg{command_line_options}};
  }

  # Make a --quiet function that turns off verbosity
  $arg{opthash}->{quiet} = sub { $arg{opthash}->{verbose} = 0; };

  unless (GetOptions($arg{opthash},
                    @default_options,
                    @extra_options,
                    )) {
    print STDERR $arg{usage};
    exit 1;
  }
  if ($arg{opthash}->{help}) {
    print $arg{usage};
    exit 0;
  }

  if ($arg{opthash}->{version}) {
    print "slack version $VERSION\n";
    exit 0;
  }

  # Get rid of the quiet handler
  delete $arg{opthash}->{quiet};

  # If we've been given a hashref, save our options there at this
  # stage, so the caller can see what was passed on the command line.
  # Unfortunately, perl has no .replace function, so we iterate.
  if (ref $arg{command_line_hash} eq 'HASH') {
    while (my ($k, $v) = each %{$arg{opthash}}) {
      $arg{command_line_hash}->{$k} = $v;
    }
  }

  # Use the default config file
  if (not defined $arg{opthash}->{config}) {
    $arg{opthash}->{config} = $DEFAULT_CONFIG_FILE;
  }

  # We need to decide whether to be verbose about reading the config file
  # Currently we just do it if global verbosity > 2
  my $verbose_config = 0;
  if (defined $arg{opthash}->{verbose}
      and $arg{opthash}->{verbose} > 2) {
    $verbose_config = 1;
  }

  # Read options from the config file, passing along the options we've
  # gotten so far
  read_config(
      file => $arg{opthash}->{config},
      opthash => $arg{opthash},
      verbose => $verbose_config,
  );

  # The "verbose" option gets compared a lot and needs to be defined
  $arg{opthash}->{verbose} ||= 0;

  # The "hostname" option is set specially if it's not defined
  if (not defined $arg{opthash}->{hostname}) {
    use Sys::Hostname;
    $arg{opthash}->{hostname} = hostname;
  }

  # We can require some options to be set
  if (ref $arg{required_options} eq 'ARRAY') {
    for my $option (@{$arg{required_options}}) {
      if (not defined $arg{opthash}->{$option}) {
        croak "Required option '$option' not given on command line or specified in config file!\n";
      }
    }
  }

  return $arg{opthash};
}

sub prompt ($) {
  my ($prompt) = @_;
  if (not defined $term) {
    require Term::ReadLine;
    $term = new Term::ReadLine 'slack'
  }

  $term->readline($prompt);
}


# Calls the callback on absolute pathnames of files in the source directory,
# and also on names of directories that don't exist in the destination
# directory (i.e. where $source/foo exists but $destination/foo does not).
sub find_files_to_install ($$$) {
  my ($source, $destination, $callback) = @_;
  return find ({
      wanted => sub {
        if (-l or not -d _) {
          # Copy all files, links, etc
          my $file = $File::Find::name;
          &$callback($file);
        } elsif (-d _) {
          # For directories, we only want to copy it if it doesn't
          # exist in the destination yet.
          my $dir = $File::Find::name;
          # We know the root directory will exist (we make it above),
          # so skip the base of the source
          (my $short_source = $source) =~ s#/$##;
          return if $dir eq $short_source;

          # Strip the $source from the path,
          # so we can build the destination dir from it.
          my $subdir = $dir;
          ($subdir =~ s#^$source##)
            or croak "sub failed: $source|$subdir";

          if (not -d "$destination/$subdir") {
            &$callback($dir);
          }
        }
      }
    },
    $source,
  );
}

# Runs rsync with the necessary redirection to its filehandles
sub wrap_rsync (@) {
  my @command = @_;
  my ($pid);

  if ($pid = fork) {
    # Parent
  } elsif (defined $pid) {
    # Child
    open(STDIN, "<", "/dev/null")
      or die "Could not redirect STDIN from /dev/null\n";
    # This redirection is necessary because rsync sends
    #   verbose output to STDOUT
    open(STDOUT, ">&STDERR")
      or die "Could not redirect STDOUT to STDERR\n";
    exec(@command);
    die "Could not exec '@command': $!\n";
  } else {
    die "Could not fork: $!\n";
  }

  my $kid = waitpid($pid, 0);
  if ($kid != $pid) {
    die "waitpid returned $kid\n";
  } elsif ($?) {
    Slack::check_system_exit(@command);
  }
}

# Runs rsync with the necessary redirection to its filehandles, but also
# returns an FH to stdin and a PID.
sub wrap_rsync_fh (@) {
  my @command = @_;
  my ($fh, $pid);

  if ($pid = open($fh, "|-")) {
    # Parent
  } elsif (defined $pid) {
    # Child
    # This redirection is necessary because rsync sends
    #   verbose output to STDOUT
    open(STDOUT, ">&STDERR")
      or die "Could not redirect STDOUT to STDERR\n";
    exec(@command);
    die "Could not exec '@command': $!\n";
  } else {
    die "Could not fork: $!\n";
  }
  return($fh, $pid);
}

1;
